lib/configuration.js

```diff
@@ -15,150 +15,335 @@
+ * directly determing whether the properties `reportUncaughtExceptions` and
+ * `key`, which can be optionally supplied in the runtime configuration, should
+ * be merged into the instance. This function also calls several specialized
+ * environmental variable checkers which not only check for the optional runtime
+ * configuration supplied values but also the processes environmental values.
+ * @memberof Configuration
+ * @private
+ * @function _gatherLocalConfiguration
+ * @returns {Undefined} - does not return anything
  */
-
+Configuration.prototype._gatherLocalConfiguration = function() {
+  if (isBoolean(this._givenConfiguration.reportUncaughtExceptions)) {
+    this._reportUncaughtExceptions = this._givenConfiguration
+      .reportUncaughtExceptions;
+  } else if (has(this._givenConfiguration, 'reportUncaughtExceptions'))  {
+    throw new Error('config.reportUncaughtExceptions must be a boolean');
+  }
+  if (isString(this._givenConfiguration.key)) {
+    this._key = this._givenConfiguration.key;
+  } else if (has(this._givenConfiguration, 'key')) {
+    throw new Error('config.key must be a string');
+  }
+};
 /**
- * Orchestrates the gathering of configuration based on submission of the
- * initConfiguration parameter. If all optional properties are given in the
- * initConfiguration parameter then the only environmental variable that will
- * be read is the NODE_ENV variable. If initConfiguration is not supplied or
- * only partially supplied then gatherConfiguration will attempt to read
- * environmental variables to supplement information as necessary.
- * @function gatherConfiguration
- * @param {ConfigurationOptions} initConfiguration - the options object
- * @returns {NormalizedConfigurationVariables} - a configuration object
- *  supplemented with environmental information as necessary
+ * The callback for `utils.getProjectId` this function is responsible for
+ * determining whether or not an error occured in the metadata transaction and,
+ * if one did not, assigning the value of the transaction (the project id) to
+ * the instance property `_projectId`.
+ * @memberof Configuration
+ * @private
+ * @function _assimilateProjectId
+ * @param {Error|Undefined} err - If applicable, the error that occured during 
+ *  the transaction
+ * @param {Null|String} projectId - The project id if successfully 
+ *  retrieved, null otherwise.
+ * @param {Function} cb - The original user callback to invoke with the project
+ *  id or error encountered during id capture
+ * @returns {Undefined} - does not return anything
  */
-function gatherConfiguration(initConfiguration) {
-
-  var projectId = null;
-  var reportUncaughtExceptions = true;
-  var serviceContext = null;
-  var key = null;
-  var shouldReportErrorsToAPI = false;
-
-  if (env.NODE_ENV === 'production') {
-
-    shouldReportErrorsToAPI = true;
+Configuration.prototype._assimilateProjectId = function(err, projectId, cb) {
```

Comment:
This function is sync, why does it accept a callback instead of returning the answer? This also looks like a candidate for inlining into the single caller since this is so simple and the term 'assimilate' doesn't intrinsically describe what this function does.